docs/architecture/adr-009-transport-agnostic-peer-abstraction.md

```diff
@@ -0,0 +1,194 @@
+        &self,
+        stream_id: &StreamId,
+    ) -> Result<(Self::Read, Self::Write), Self::Error>;
+    fn public_key(&self) -> PublicKey;
+    fn remote_addr(&self) -> SocketAddr;
+}
+```
+
+### Peer
+
+Given a correct implementation of a `Transport` and its `Connection` newly
+established ones will be wrapped with a `Peer`. Which is in charge of setting
+up the correct streams on the `Connection` and multiplex messages - incoming
+and outgoing alike - efficiently. It's also an attempt to enforce
+correct-by-construction constraints on the state machine of the peer. To avoid
+misuse or unexpected transitions. The only way to construct is, is from an
+existing connection which gives the caller a connected peer. When invoking run
+on that one a fully function peer is "returned". Therefore the states look
+like: `Connected -> Running -> Stopped`.
+
+``` rust
+impl<Conn> Peer<Connected<Conn>>
+where
+    Conn: Connection,
+{
+    pub fn run(self, stream_ids: Vec<StreamId>) -> Result<Peer<Running<Conn>>> {
+        // ...
+    }
+
+    fn stop(self) -> Result<Peer<Stopped>> {
+        // ...
+    }
+}
+
+impl<Conn> Peer<Running<Conn>>
+where
+    Conn: Connection,
+{
+    pub fn send(&self, message: message::Send) -> Result<()> {
+        // ...
+    }
+
+    pub fn stop(self) -> Result<Peer<Stopped>> {
+        // ...
+    }
+}
+```
+
+While sending messages is done through a method on a running peer, getting hold
+of incoming messages can be achieved by draining the `Receiver` part of the
+running state.
+
+
+### Supervisor
+
+The `Supervisor` is the main entry point to the p2p package giving higher-level
+components access to a unified stream of peer events and messages as well as
+the ability to control peer lifecycle (connect, disconnect, etc.).
+
+``` rust
+pub enum Command {
+    Accept,
+    Connect(SocketAddr),
+    Disconnect(node::Id),
+    Msg(node::Id, message::Send),
+}
+
+pub enum Event {
+    Connected(node::Id, Direction),
+    Disconnected(node::Id, Report),
+    Message(node::Id, message::Receive),
+    Upgraded(node::Id),
+    UpgradeFailed(node::Id, Report),
+}
+
+impl Supervisor {
+    pub fn run<T>(transport: T) -> Result<Self>
+    where
+        T: transport::Transport + Send + 'static,
+    {
+        // ...
+    }
+
+    pub fn recv(&self) -> Result<Event> {
+        // ...
+    }
+
+    pub fn command(&self, cmd: Command) -> Result<()> {
+        // ...
+    }
+}
+```
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+* Unified way to bootstrap and integrate transports
+* Potential for connecting different wire transports in the same process
+* Rest of the domain is simply concerned with `PeerId`s as identity
+
+### Negative
+
+* Significant set of abstractions need to be satisfied for a new transport
+  implementation
+* Non-stream based transports need to be fitted into this model
+
+### Neutral
+
+## Open Questions
+
+* Should serialization be transport specific to allow for optimisations on the
```

Comment:
NOTE: tendermint-go has decided on unified serialization (protobuf) across all transports for the sake of simplicity.

Reply:
Great, thanks for the context. If this is indeed the vision @brapse @greg-szabo @erikgrinaker? We should stick to the plan that connections provide byte streams and we maintain a single implementation of the serialisation.

Personally I would benefit from a brush up on how currently the byte layout looks on the wire with TCP connections. Are we still byte-prefixing to determine the `channel`?